File: HoverFillButton/hoverfillbutton.cpp
#include "hoverfillbutton.h"

HoverFillButton::HoverFillButton(QWidget *parent, AnimationType animationType)
    : HoverFillButtonBase::HoverFillButtonBase(parent), animation_type{animationType}
{
    initializeMemberVariable();
}
HoverFillButton::HoverFillButton(const QString &text, QWidget *parent, AnimationType animationType)
    : HoverFillButtonBase::HoverFillButtonBase(text, parent), animation_type{animationType}
{
    initializeMemberVariable();
}

HoverFillButton::HoverFillButton(const QIcon &icon, const QString &text, QWidget *parent, AnimationType animationType)
    : HoverFillButtonBase::HoverFillButtonBase(icon, text, parent), animation_type{animationType}
{
    initializeMemberVariable();
}

void HoverFillButton::setAnimationType(AnimationType animationType)
{
    animation_type = animationType;
    initializeMemberVariable();
}

void HoverFillButton::paintValueReduction()
{
    HoverFillButtonBase::paintValueReduction();
    update();
}

void HoverFillButton::paintValueIncrease()
{
    HoverFillButtonBase::paintValueIncrease();
    update();
}

void HoverFillButton::enterEvent(QEnterEvent *event)
{
    HoverFillButtonBase::enterEvent(event);

    connect(timer, &QTimer::timeout, this, &HoverFillButton::paintValueIncrease); // 定时器触发，半径增大
    switch (animation_type)
    {
    case AnimationType::CircularFill:    // 进入点圆形填充
        mouse_point = event->position(); // 记录鼠标进入坐标
        break;
    default:
        break;
    }
    timer->start();
}

void HoverFillButton::leaveEvent(QEvent *event)
{
    HoverFillButtonBase::leaveEvent(event);

    connect(timer, &QTimer::timeout, this, &HoverFillButton::paintValueReduction);
    switch (animation_type)
    {
    case AnimationType::CircularFill: // 进入点圆形填充
        mouse_point = mapFromGlobal(QCursor::pos());
        break;
    default:
        break;
    }
    timer->start();
}

void HoverFillButton::resizeEvent(QResizeEvent *event)
{
    HoverFillButtonBase::resizeEvent(event);
    initializeMemberVariable();
}

void HoverFillButton::paintEvent(QPaintEvent *event)
{
    HoverFillButtonBase::paintEvent(event);
    QPainter painter(this);
    painter.setRenderHints(QPainter::Antialiasing | QPainter::TextAntialiasing); // 抗锯齿和文字走样
    painter.setBrush(fill_brush);                                        // 设置填充颜色
    painter.setPen(Qt::NoPen);                                                   // 设置无边框
    painter.setClipPath(path);                                                   // 绘制设置路径
    painter.setFont(font());                                                     // 设置字体
    switch (animation_type)
    {
    case AnimationType::CircularFill: // 进入点圆形填充
        if (!mouse_point.isNull())
            painter.drawEllipse(mouse_point, paint_value, paint_value); // 从鼠标进入点开始画圆
        break;
    case AnimationType::CrossFill: // 左右两个小圆交叉填充
        painter.drawEllipse(QPoint(0, 0), paint_value, paint_value);
        painter.drawEllipse(QPoint(width(), height()), paint_value, paint_value);
        break;
    case AnimationType::DiagonalRectangle:                                              // 斜着的矩形填充
        painter.save();                                                                 // 保存状态
        painter.translate(width() / 2, height() / 2);                                   // 将坐标原点移动到按钮中心
        painter.rotate(45);                                                             // 将坐标系旋转45度
        painter.drawRect(QRectF(-paint_value / 2, -width() / 2, paint_value, width())); // 绘制矩形
        painter.restore();                                                              // 恢复状态
        break;
    case AnimationType::BottomCircle: // 底部圆形填充
        painter.save();
        painter.translate(width() / 2, int(height() * 1.5));
        painter.drawEllipse(QPoint(0, 0), paint_value, paint_value);
        painter.restore();
        break;
    case AnimationType::LeftRectangle: // 左侧矩形填充
        painter.save();
        if (is_cursor_inside) // 光标在内部,从左到右绘制矩形
            painter.drawRect(QRectF(0, 0, paint_value, height()));
        else // 坐标系翻转,从右到左绘画矩形
        {
            painter.translate(width(), height());
            painter.rotate(180);
            painter.drawRect(QRectF(0, 0, paint_value, height()));
        }
        painter.restore();
        break;
    }
    if (paint_value > 0)
    {
        painter.setPen(hover_text_color);                  // 设置画笔颜色
        painter.drawText(rect(), Qt::AlignCenter, text()); // 绘制文字
    }
}

void HoverFillButton::initializeMemberVariable()
{
    switch (animation_type)
    {
    case AnimationType::CircularFill:                                      // 进入点,圆形填充
        max_painte_value = qSqrt(width() * width() + height() * height()); // 计算最大半径
        break;
    case AnimationType::CrossFill: // 左右两个小圆交叉填充
        max_painte_value = qPow(qPow(width(), 2) + qPow(height(), 2), 0.5);
        break;
    case AnimationType::DiagonalRectangle: // 斜着的矩形填充
        max_painte_value = qPow(qPow(width(), 2) + qPow(height(), 2), 0.5);
        break;
    case AnimationType::BottomCircle: // 底部圆形填充
        max_painte_value = qPow(qPow(width(), 2) + qPow(height(), 2), 0.5);
        break;
    case AnimationType::LeftRectangle: // 左侧矩形填充
        max_painte_value = qPow(qPow(width(), 2) + qPow(height(), 2), 0.5);
        break;
    }
}

File: Archive scrap code/PySide6/HoverFillButton/HoverFillButton.py
import math
from PySide6.QtCore import Slot, QPointF, Qt, QPoint, QRectF
from PySide6.QtGui import QCursor, QPainter
from PySide6.QtWidgets import QWidget
from HoverFillButtonBase import HoverFillButtonBase

qSqrt = math.sqrt
qPow = math.pow


class HoverFillButton(HoverFillButtonBase):
    def __init__(self, text: str = None, parent: QWidget = None):
        super().__init__(text=text, parent=parent)
        self.animationType = self.AnimationType.CircularFill
        self.mouse_point = QPointF()
        self.initializeMemberVariable()

    def setAnimationType(self, animationType: HoverFillButtonBase.AnimationType):
        self.animationType = animationType
        self.initializeMemberVariable()

    @Slot()
    def paintValueReduction(self):
        super().paintValueReduction()
        self.update()

    @Slot()
    def paintValueIncrease(self):
        super().paintValueIncrease()
        self.update()

    def enterEvent(self, event):
        super().enterEvent(event)
        self.timer.timeout.connect(self.paintValueIncrease)
        if self.animationType == HoverFillButtonBase.AnimationType.CircularFill:
            self.mouse_point = self.mapFromGlobal(QCursor.pos())
        self.timer.start()

    def leaveEvent(self, event):
        super().leaveEvent(event)
        self.timer.timeout.connect(self.paintValueReduction)
        if self.animationType == HoverFillButtonBase.AnimationType.CircularFill:
            self.mouse_point = self.mapFromGlobal(QCursor.pos())
        self.timer.start()

    def resizeEvent(self, event):
        super().resizeEvent(event)
        self.initializeMemberVariable()

    def paintEvent(self, event):
        super().paintEvent(event)
        painter = QPainter(self)
        painter.setRenderHint(QPainter.RenderHint.Antialiasing | QPainter.RenderHint.TextAntialiasing)
        painter.setBrush(self.fill_brush)
        painter.setPen(Qt.PenStyle.NoPen)
        painter.setClipPath(self.path)
        painter.setFont(self.font())

        if self.animationType == HoverFillButtonBase.AnimationType.CircularFill:
            if not self.mouse_point.isNull():
                painter.drawEllipse(self.mouse_point, self.paint_value, self.paint_value)
        if self.animationType == HoverFillButtonBase.AnimationType.CrossFill:
            painter.drawEllipse(QPoint(0, 0), self.paint_value, self.paint_value)
            painter.drawEllipse(QPoint(self.width(), self.height()), self.paint_value, self.paint_value)
        if self.animationType == HoverFillButtonBase.AnimationType.DiagonalRectangle:
            painter.save()
            painter.translate(self.width() / 2, self.height() / 2)
            painter.rotate(45)
            painter.drawRect(QRectF(-self.paint_value / 2, -self.width() / 2, self.paint_value, self.width()))
            painter.restore()

        if self.animationType == HoverFillButtonBase.AnimationType.BottomCircle:
            painter.save()
            painter.translate(self.width() / 2, int(self.height() * 1.5))
            painter.drawEllipse(QPoint(0, 0), self.paint_value, self.paint_value)
            painter.restore()

        if self.animationType == HoverFillButtonBase.AnimationType.LeftRectangle:
            painter.save()
            if self.is_cursor_inside:
                painter.drawRect(QRectF(0, 0, self.paint_value, self.height()))
            else:
                painter.translate(self.width(), self.height())
                painter.rotate(180)
                painter.drawRect(QRectF(0, 0, self.paint_value, self.height()))
            painter.restore()
        if self.paint_value > 0:
            painter.setPen(self.hover_text_color)
            painter.drawText(self.rect(), Qt.AlignmentFlag.AlignCenter, self.text())

    def initializeMemberVariable(self):
        if self.animationType == HoverFillButtonBase.AnimationType.CircularFill:
            self.max_painted_value = qSqrt(self.width() * self.width() + self.height() * self.height())
            return
        else:
            self.max_painted_value = int(qPow(qPow(self.width(), 2) + qPow(self.height(), 2), 0.5))
            return
